app/src/main/java/ee/ioc/phon/android/speechutils/AbstractAudioRecorder.java

```diff
@@ -128,20 +152,71 @@ protected int getStatus(int numOfBytes, int len) {
+     *                       passed the end of the buffer and returned to the beginning
+     * @param numOfBytesRead - in the reading process
+     * @return true/false according to the above logic
+     */
+    private boolean isConsumePointerCrossed(boolean reachedTheEndOfRecordingBuffer, int numOfBytesRead) {
+        return numOfBytesRead > 0 &&
+               mRecordingBufferIsFullWithData &&
+               isRecordedSessionSameAsConsumedSession() &&
+               ((mRecordedLength - numOfBytesRead < mConsumedLength && mRecordedLength >= mConsumedLength) ||
+                (reachedTheEndOfRecordingBuffer && mConsumedLength < mRecordedLength));
+    }
+
+    /**
      * Copy data from the given recorder into the given buffer, and append to the complete recording.
      * public int read (byte[] audioData, int offsetInBytes, int sizeInBytes)
      */
     protected int read(SpeechRecord recorder, byte[] buffer) {
         int len = buffer.length;
         int numOfBytes = recorder.read(buffer, 0, len);
         int status = getStatus(numOfBytes, len);
-        if (status == 0) {
-            // arraycopy(Object src, int srcPos, Object dest, int destPos, int length)
-            // numOfBytes <= len, typically == len, but at the end of the recording can be < len.
-            System.arraycopy(buffer, 0, mRecording, mRecordedLength, numOfBytes);
-            mRecordedLength += len;
+        boolean reachedTheEndOfRecordingBuffer = false;
+        // if we need to keep on listening, when reaching the end of the recorded buffer,
+        // continue to write from the beginning. thus, we have a cyclic buffer
+        if (mAlwaysListen && status == -300) {
+            reachedTheEndOfRecordingBuffer = true;
+            status = 0;
+            // for use when consuming the recorded buffer, the buffer is now in it's cyclic phase
+            if (!mRecordingBufferIsFullWithData)
+                mRecordingBufferIsFullWithData = true;
         }
-        return status;
+
+        if (status == 0 && numOfBytes >= 0) {
+            if (!reachedTheEndOfRecordingBuffer) {
+                // arraycopy(Object src, int srcPos, Object dest, int destPos, int length)
+                // numOfBytes <= len, typically == len, but at the end of the recording can be < len.
+                System.arraycopy(buffer, 0, mRecording, mRecordedLength, numOfBytes);
+                mRecordedLength += numOfBytes;
+            }
+            else {
+                int numOfBytesBeforeCyclic = mRecording.length - mRecordedLength;
+                System.arraycopy(buffer, 0, mRecording, mRecordedLength, numOfBytesBeforeCyclic);
+                System.arraycopy(buffer, numOfBytesBeforeCyclic, mRecording, 0, numOfBytes - numOfBytesBeforeCyclic);
+
+                mRecordedLength = numOfBytes - numOfBytesBeforeCyclic;
+            }
+
+            // increment the recorded session id in case that the consume pointer was crossed
```

Comment:
change the recording session id in case that the recording pointer just passed the consuming pointer